### ISSUE TITLE
Compilation problem with symengine

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#15457]
Something was merged over the past couple of days that broke compilation for me when using symengine:
```
/home/bangerth/p/deal.II/1/dealii/source/differentiation/sd/symengine_number_types.cc: In constructor ‘dealii::Differentiation::SD::Expression::Expression(const std::vector<std::pair<dealii::Differentiation::SD::Expression, dealii::Differentiation::SD::Expression> >&, const dealii::Differentiation::SD::Expression&)’:
/home/bangerth/p/deal.II/1/dealii/source/differentiation/sd/symengine_number_types.cc:107:34: error: cannot bind rvalue reference of type ‘SymEngine::PiecewiseVec&&’ {aka ‘std::vector<std::pair<SymEngine::RCP<const SymEngine::Basic>, SymEngine::RCP<const SymEngine::Boolean> > >&&’} to lvalue of type ‘SymEngine::PiecewiseVec’ {aka ‘std::vector<std::pair<SymEngine::RCP<const SymEngine::Basic>, SymEngine::RCP<const SymEngine::Boolean> > >’}
  107 |       expression = SE::piecewise(piecewise_function);
      |                                  ^~~~~~~~~~~~~~~~~~
In file included from /home/bangerth/p/deal.II/1/dealii/include/deal.II/differentiation/sd/symengine_number_types.h:32,
                 from /home/bangerth/p/deal.II/1/dealii/source/differentiation/sd/symengine_number_types.cc:22:
/home/bangerth/bin/candi-install/symengine-0.8.1/include/symengine/logic.h:101:50: note:   initializing argument 1 of ‘SymEngine::RCP<const SymEngine::Basic> SymEngine::piecewise(SymEngine::PiecewiseVec&&)’
  101 | inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
      |                                   ~~~~~~~~~~~~~~~^~~
```
We ought to fix this before the release :-)

**Comment by bangerth** [comment]
It was this part of #15422: https://github.com/dealii/dealii/pull/15422/files#diff-2c3b4f7640b5ae1a3b05deecf96447b2e09a22e8368ac24271b692d7eb6b29d1

**Comment by bangerth** [comment]
Fixed by #15458.
